### DISCUSSION TITLE
How to handle state in a cloud native environment

### DISCUSSION THREAD

**Discussion by Vad1mo** [tetra-framework/tetra#10] [Q&A] [answered]
I read that Tetra is using Pickle to store a state between requests. I was wondering how would Tetra work in a Cloud Native environment with containers coming and going and where you usually have at least 2 instances running to have some sort of resilience. 

Is it possible to configure pickle to share the state across multiple instance, is is possible to resume after a container crash?

**Comment by samwillis** [comment, marked as answer]
Tetra is prefect for that sort of deployment as although the components are "statefull", that state is not persisted on the server. They are actually "paused", serialised with [Pickle (Pythons built in serialization api)](https://docs.python.org/3/library/pickle.html), encrypted+signed (for security), and that is then sent to the browser. The browser then sends the encrypted state back to the server each time it calls a public method, receiving an updated encrypted state in return. No state is persisted on the server at all - perfect for multi server deployments or "serverless".

You can read details of how it works here: https://www.tetraframework.com/docs/state-security

**Reply by Vad1mo** [reply]
ahh ok, got it, have seen the docs but jumped over that particular sentence, thank you

**Reply by jaycle** [reply]
This is really a great idea. I, too, didn't notice in the docs at first how the architecture of state was working. Maybe the docs could put **State Security** in just a **State** section?
